### PR TITLE
RP-366 Create Sandbox API/Service

### PR DESCRIPTION
**Pull request by vgonzalez9**
Moving the AggregateReporting configuration files to rdw_reporting_common
first commit with successful IT build. There will be another one with all of the SandboxService changes.

This was done with the latest code, and because automatic and patched merges failed.
The second commit is the Sandbox API changes.

### CODE REVIEW

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/model/ApplicationSandboxConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import org.opentestsystem.rdw.multitenant.TenantProperties;
+import org.opentestsystem.rdw.reporting.common.configuration.AggregateReportingPropertiesTenant;
+import org.opentestsystem.rdw.reporting.common.configuration.ReportingSystemPropertiesImpl;
+
+import java.util.Set;
+
+public class ApplicationSandboxConfiguration {
+    private TenantProperties.Tenant tenant; // Tenant used to create sandbox, and where sandbox exists
+    private Set<DataSet> dataSets; // collection of dataSets preconfigured for this tenant
+    private ReportingSystemPropertiesImpl reporting; // default Reporting system properties
```
Comment (malaffoon) [inline admin-service/src/main/java/org/opentestsystem/rdw/admin/model/ApplicationSandboxConfiguration.java]: please put these comments in the javadocs for the getters

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/model/ApplicationSandboxConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package org.opentestsystem.rdw.admin.model;
+
+import org.opentestsystem.rdw.multitenant.TenantProperties;
+import org.opentestsystem.rdw.reporting.common.configuration.AggregateReportingPropertiesTenant;
+import org.opentestsystem.rdw.reporting.common.configuration.ReportingSystemPropertiesImpl;
+
+import java.util.Set;
+
+public class ApplicationSandboxConfiguration {
```
Comment (malaffoon) [inline admin-service/src/main/java/org/opentestsystem/rdw/admin/model/ApplicationSandboxConfiguration.java]: overview javadoc would be helpful

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/model/Sandbox.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    public void setId(final String id) {
+        this.id = id;
+    }
+
+    public String getKey() {
+        return key;
+    }
+
+    public void setKey(final String key) {
```
Comment (malaffoon) [inline admin-service/src/main/java/org/opentestsystem/rdw/admin/model/Sandbox.java]: I'm assuming `key` here is the sandbox equivalent of tenant key?

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/model/DataSet.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+    public String getLabel() {
+        return label;
+    }
+
+    public void setLabel(final String label) {
+        this.label = label;
+    }
+
+    public String getKey() {
```
Comment (malaffoon) [inline admin-service/src/main/java/org/opentestsystem/rdw/admin/model/DataSet.java]: What is `key` here? I'm guessing it is an identifier of some kind so i'd much prefer `id`.

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/model/Sandbox.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
+package org.opentestsystem.rdw.admin.model;
+
+public class Sandbox {
+    private String id;
+    private String key;
+    private String name;
+    private String description;
```
Comment (malaffoon) [inline admin-service/src/main/java/org/opentestsystem/rdw/admin/model/Sandbox.java]: Is there any way we can share the common Tenant class here? Perhaps we can move `Tenant` out of `TenantProperties` and then use it as a base class here? That way it becomes much more obvious that Sandbox isA Tenant.

Comment (ftdholbrook) [reply admin-service/src/main/java/org/opentestsystem/rdw/admin/model/Sandbox.java]: I like that idea.

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/repository/SandboxConfigurationRepository.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
+package org.opentestsystem.rdw.admin.repository;
+
+import org.opentestsystem.rdw.admin.model.SandboxConfiguration;
+
+import java.util.Set;
+
+public interface SandboxConfigurationRepository {
+    /**
+     * @return the sandbox default configuration and the list of existing sandboxes
```
Comment (malaffoon) [inline admin-service/src/main/java/org/opentestsystem/rdw/admin/repository/SandboxConfigurationRepository.java]: Bit confusing, is the default configuration just one of the entries in the set?

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/repository/SandboxConfigurationRepository.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    Set<SandboxConfiguration> getAll();
+
+    /**
+     * @param sandboxConfiguration to create
+     * @return the newly added sandboxConfiguration
+     */
+    SandboxConfiguration add(SandboxConfiguration sandboxConfiguration);
```
Comment (malaffoon) [inline admin-service/src/main/java/org/opentestsystem/rdw/admin/repository/SandboxConfigurationRepository.java]: `create` would be more consistent

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/repository/SandboxConfigurationRepository.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * @param key of the Sandbox to reset its DataSet
+     */
+    void reset(String key);
+
+    /**
+     * @param key of the sandbox to check
+     * @return true if the sandbox key exists
+     */
+    Boolean sandboxExists(String key);
```
Comment (malaffoon) [inline admin-service/src/main/java/org/opentestsystem/rdw/admin/repository/SandboxConfigurationRepository.java]: This should be `boolean exists(String key)`: it is a method on the SandboxRepository so `sandbox` is redundant, and the return will always be a value so using the primitive `boolean` is appropriate.

**Comment by malaffoon** [comment]
@vgonzalez9, i merged develop (with my changes to common) into this branch ... didn't want you to have to deal with conflicts.

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/model/Sandbox.java (hunk 5 above)
Comment (vgonzalez9) [reply admin-service/src/main/java/org/opentestsystem/rdw/admin/model/Sandbox.java]: Will fix.

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/repository/SandboxConfigurationRepository.java (hunk 6 above)
Comment (vgonzalez9) [reply admin-service/src/main/java/org/opentestsystem/rdw/admin/repository/SandboxConfigurationRepository.java]: You are correct, it really just the defaults of a tenant.
will clarify.

File: admin-service/src/main/java/org/opentestsystem/rdw/admin/model/Sandbox.java (hunk 5 above)
Comment (malaffoon) [reply admin-service/src/main/java/org/opentestsystem/rdw/admin/model/Sandbox.java]: I've separated out Tenant so you don't have to worry about that part.

**Pull request merged**